larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc

```diff
@@ -0,0 +1,237 @@
+/**
+ *  @file   larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc
+ *
+ *  @brief  Implementation of the lar discrete probability vector class
+ *
+ *  $Log: $
+ */
+
+#include "larpandoracontent/LArObjects/LArDiscreteProbabilityVector.h"
+
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+
+namespace lar_content
+{
+
+template <typename TX, typename TY>
+DiscreteProbabilityVector::DiscreteProbabilityVector(const InputData<TX, TY> &inputData, const TX xUpperBound, 
+    const bool useWidths) :
+    m_xUpperBound(static_cast<float>(xUpperBound)),
+    m_useWidths(useWidths),
+    m_discreteProbabilityData(this->InitialiseDiscreteProbabilityData(inputData))
+{
+    this->VerifyCompleteData();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+DiscreteProbabilityVector::DiscreteProbabilityVector(const DiscreteProbabilityVector &discreteProbabilityVector,
+    std::mt19937 &randomNumberGenerator) :
+    m_xUpperBound(discreteProbabilityVector.m_xUpperBound),
+    m_useWidths(discreteProbabilityVector.m_useWidths),
+    m_discreteProbabilityData(this->RandomiseDiscreteProbabilityData(discreteProbabilityVector, randomNumberGenerator))
+{
+    this->VerifyCompleteData();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+DiscreteProbabilityVector::DiscreteProbabilityVector(const DiscreteProbabilityVector &discreteProbabilityVector,
+    const ResamplingPoints &resamplingPoints) :
+    m_xUpperBound(discreteProbabilityVector.m_xUpperBound),
+    m_useWidths(discreteProbabilityVector.m_useWidths),
+    m_discreteProbabilityData(this->ResampleDiscreteProbabilityData(discreteProbabilityVector, resamplingPoints))
+{
+    this->VerifyCompleteData();
+}
+
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float DiscreteProbabilityVector::EvaluateCumulativeProbability(const float x) const
+{
+    if (x - m_discreteProbabilityData.back().GetX() > std::numeric_limits<float>::epsilon())
+        return 1.f;
+
+    if (x - m_discreteProbabilityData.front().GetX() < std::numeric_limits<float>::epsilon())
+        return 0.f;
+
+    for (size_t iDatum = 1; iDatum < m_discreteProbabilityData.size(); ++iDatum)
+    {
+        if (x - m_discreteProbabilityData.at(iDatum).GetX() > std::numeric_limits<float>::epsilon())
+            continue;
+
+        const float xLow(m_discreteProbabilityData.at(iDatum-1).GetX());
```

Comment:
Spaces around operators, throughout.

Reply:
Mostly yes.  When there are +- and * on the same line I tend to try to keep no spacing around * as it helps the mind employ bidmas.
I'll leave this one open if you want to come back to it.